### PR TITLE
Add example of Pearson correlation with callbacks

### PR DESCRIPTION
**Pull request by pbarbarant**
Small PR to add a miscellaneous example to show how to monitor the Pearson correlation during training.

### CODE REVIEW

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+# %%
+# We then define a callback function that computes the
+# Pearson correlation between the transformed features and
+# the target features at each BCD iteration.
+
+# Initialize the list of Pearson correlations with the original
+# untansformed features
+corr_bcd_steps = [
+    pearson_corr(source_features_normalized, target_features_normalized)
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:125]: I don't think this is a fair initialization: you should probably compute the pearson correlation between source features transported on a target anatomy using our init transport plan (or at least between shuffled source and target features).
What you are computing here is more comparable to a baseline I think (pearson correlation between features of voxels which were projected on a template anatomy).

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    verbose=True,
+)
+
+# %%
+# We finally plot the evolution of the Pearson correlation and
+# the training loss at each BCD iteration. We can see that the
+# Pearson correlation dips at the first iteration, but quickly
+# increases and stabilizes at the second iteration. This
+# is due to the initialization of the transport plan, which
+# is filled with ones.
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:185]: This comment will probably no longer be true after you make updates related to my previous point

**Review by alexisthual — COMMENTED** [review]
Nice start, your flow is easy to read :)
I have concerns with your correlation init though and think it should be revised.

In a follow-up PR, I think we could add util functions which create init transport plans (I have 2 inits in mind, the "entropic" one which yields an OT plan with 1s everywhere - it's the one we have now - and a "random" one which the solution to a 1D wasserstein problem between randomly sampled values in [0, 1] - we might need to add POT to our dependencies).

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+
+from functools import partial
+
+from nilearn import datasets, image
+from scipy.spatial import distance_matrix
+from fugw.mappings import FUGW
+from fugw.utils import _make_tensor
+from rich.console import Console
```
[the remark this answers is not in the document]
Comment (pbarbarant) [reply examples/02_miscellaneous/plot_1_pearson_correlation.py:24]: Just tried it, works well, but I agree we should do it in another fixing PR.

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 4]
```diff
@@ -4,24 +4,24 @@
 ==============================================================
 
 In this example, we use a callback function to monitor the Pearson correlation
-between the transformed features and the target features at each iteration of
+between transformed and target features at each iteration of
 the block-coordinate descent (BCD) algorithm.
 This can be useful to detect numerical errors, or to check that the
-mapping is fitting the data correctly.
+mapping is not over-fitting training data.
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:10]:
```suggestion
mapping is not over-fitting training data.
Note that in this example, we don't use a proper validation dataset.
Instead, we simply compute more metrics on our training dataset.
```

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 5]
```diff
@@ -97,32 +97,51 @@
...
-# the target features at each BCD iteration.
+# Pearson correlation between transformed features and
+# target features at each BCD iteration.
+
+
+# Initialize the transport plan with ones and normalize it
+init_plan = torch.ones(
+    (source_features_normalized.shape[1], source_features_normalized.shape[1])
+).to(float)
+init_plan_normalized = init_plan / init_plan.sum()
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:135]: Great!
Now, I think you should use this plan when fitting your mapping (simply add `init_plan=init_plan` in `mapping.fit()`.
You and I know that what fugw does under the hood is to instantiate a plan equal to the one you just created, but this might change in the future.

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 6]
```diff
@@ -177,12 +189,10 @@ def correlation_callback(
...
-# We finally plot the evolution of the Pearson correlation and
-# the training loss at each BCD iteration. We can see that the
-# Pearson correlation dips at the first iteration, but quickly
-# increases and stabilizes at the second iteration. This
-# is due to the initialization of the transport plan, which
-# is filled with ones.
+# The Pearson correlation and training loss evolution are then plotted for
+# each BCD iteration. As we begin over-fitting the training data, the
+# correlation rises during the initial iterations before gradually
+# falling down.
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:195]: I don't think you can make such a claim using training data only.
In particular, I think you can only deduce that your algorithm is over-fitting by looking at metrics computed on validation data.
However, what we can say here is that a lower fugw loss does not always result in higher functional correlations.

**Review by alexisthual — COMMENTED** [review]
I have minor comments but this PR is essentially good to go, good job! :)

Feel free to keep this for later (or never haha), but a nice addition could be to have an actual validation set here (ie some contrast maps which are not used to fit our mapping), which would allow you to plot the dynamics of functional correlation on a validation set, and nicely conclude this example.

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        x = torch.tensor(x).to(torch.float32)
+        y = torch.tensor(y).to(torch.float32)
+
+    # Compute the transformed features
+    x_transformed = (
+        (plan @ x.T / plan.sum(dim=1).reshape(-1, 1)).T.detach().cpu()
+    )
+
+    vx = x_transformed - torch.mean(x_transformed)
+    vy = y - torch.mean(y)
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:117]: Actually, `torch.mean(x_transformed)` is not what you want here: you need to compute the mean along a specific axis.

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+    corr = (
+        torch.sum(vx * vy)
+        / (torch.sqrt(torch.sum(vx**2) * torch.sum(vy**2)))
+    ).numpy()
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:122]: Same here, you should sum along a specific axis ; `corr` should not simply be a scalar but an array

**Review by alexisthual — COMMENTED** [review]
OT transformation is not correct :scream:

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+)
+ax2.fill_between(
+    mapping.loss_steps[: len(corr_bcd_steps)],
+    mean_corr - std_corr,
+    mean_corr + std_corr,
+    color=color,
+    alpha=0.2,
+    label="1st standard deviation",
+)
+ax2.set_ylim(0, 1)
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:317]: 1 validation contrast is missing from your plot because it has negative correlation
```suggestion
ax2.set_ylim(-0.5, 1)
```

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+    color="blue",
+    label="Average correlation",
+)
+ax2.fill_between(
+    mapping.loss_steps[: len(corr_bcd_steps)],
+    mean_corr - std_corr,
+    mean_corr + std_corr,
+    color=color,
+    alpha=0.2,
+    label="1st standard deviation",
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:315]:
```suggestion
    label="Standard deviation",
```

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+        alpha=0.5,
+        linestyle="dashed",
+    )
+ax2.set_label("Pearson correlation")
+
+ax2.plot(
+    mapping.loss_steps[: len(corr_bcd_steps)],
+    mean_corr,
+    color="blue",
+    label="Average correlation",
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:307]:
```suggestion
    label="Average across-voxels correlation",
```

File: examples/02_miscellaneous/plot_1_pearson_correlation.py [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+color = "tab:blue"
+ax2.set_ylabel("Pearson correlation", color=color)
+for i in range(len(source_features_normalized[n_training_contrasts:])):
+    ax2.plot(
+        mapping.loss_steps[: len(corr_bcd_steps)],
+        corr_bcd_steps[:, i],
+        color=color,
+        alpha=0.5,
+        linestyle="dashed",
+    )
```
Comment (alexisthual) [inline examples/02_miscellaneous/plot_1_pearson_correlation.py:300]: It would be nice if these lines had their pattern explained in your legend

**Review by alexisthual — APPROVED** [review]
LGTM!
Minor suggestions, feel free to leave or integrate them.

**Comment by pbarbarant** [comment]
> > I added validation data to improve the final aspect of the graph with the last 4 commits. The final plot is much nicer and easier to understand compared to the previous one which _**spiked**_ after the first BCD iteration.
> 
> Empirically, on other datasets, I see this spike on training and validation sets. Do you see another explanation as to why your recent changes seem to have affected this?

I reshuffled the data between my training and validation set. Some contrasts exhibit this spike, others don't. I feel like it highly depends on the diversity of the training data used to fit the mapping.
